Mathlib/RingTheory/DividedPowers/Basic.lean

```diff
@@ -0,0 +1,497 @@
+Authors: Antoine Chambert-Loir, María-Inés de Frutos—Fernández
+-/
+
+import Mathlib.RingTheory.PowerSeries.Basic
+import Mathlib.Combinatorics.Enumerative.Bell
+import Mathlib.Data.Nat.Choose.Multinomial
+
+/-! # Divided powers
+
+Let `A` be a commutative (semi)ring and `I` be an ideal of `A`.
+A *divided power* structure on `I` is the datum of operations `a n ↦ dpow a n`
+satisfying relations that model the intuitive formula `dpow n a = a ^ n / n.factorial` and
+collected by the structure `DividedPowers`. The list of axioms is embedded in the structure:
+To avoid coercions, we rather consider `DividedPowers.dpow : ℕ → A → A`, extended by 0.
+
+* `DividedPowers.dpow_null` asserts that `dpow n x = 0` for `x ∉ I`
+
+* `DividedPowers.dpow_mem` : `dpow n x ∈ I` for `n ≠ 0`
+For `x y : A` and `m n : ℕ` such that `x ∈ I` and `y ∈ I`, one has
+* `DividedPowers.dpow_zero` : `dpow 0 x = 1`
+* `DividedPowers.dpow_one` : `dpow 1 x = 1`
+* `DividedPowers.dpow_add` : `dpow n (x + y) =
+(antidiagonal n).sum fun k => dpow k.1 x * dpow k.2 y`,
+this is the binomial theorem without binomial coefficients.
+* `DividedPowers.dpow_mul`: `dpow n (a * x) = a ^ n * dpow n x`
+* `DividedPowers.mul_dpow` : `dpow m x * dpow n x = choose (m + n) m * dpow (m + n) x`
+* `DividedPowers.dpow_comp` : `dpow m (dpow n x) = uniformBell m n * dpow (m * n) x`
+
+* `DividedPowers.dividedPowersBot` : the trivial divided powers structure on the zero ideal
+
+* `DividedPowers.prod_dpow`: a product of divided powers is a multinomial coefficients
+times a divided power
+
+* `DividedPowers.dpow_sum`: the multinomial theorem for divided powers,
+without multinomial coefficients.
+
+* `DividedPowers.ofRingEquiv`: transfer divided powers along `RingEquiv`
+
+* `DividedPowers.equiv`: the equivalence `DividedPowers I ≃ DividedPowers J`,
+for `e : R ≃+* S`, and `I : Ideal R`,  `J : Ideal S` such that `I.map e = J`
+
+* `DividedPowers.exp`: the power series `Σ (dpow n a) X ^n`
+
+* `DividedPowers.exp_add`: its multiplicativity
+
+## References
+
+* [P. Berthelot (1974), *Cohomologie cristalline des schémas de
+caractéristique $p$ > 0*][Berthelot-1974]
+
+* [P. Berthelot and A. Ogus (1978), *Notes on crystalline
+cohomology*][BerthelotOgus-1978]
+
+* [N. Roby (1963), *Lois polynomes et lois formelles en théorie des
+modules*][Roby-1963]
+
+* [N. Roby (1965), *Les algèbres à puissances dividées*][Roby-1965]
+
+## Discussion
+
+* In practice, one often has a single such structure to handle on a given ideal,
+but several ideals of the same ring might be considered.
+Without any explicit mention of the ideal, it is not clear whether such structures
+should be provided as instances.
+
+* We do not provide any notation such as `a ^[n]` for `dpow a n`.
+
+-/
+
+
+section DividedPowersDefinition
+/- ## Definition of divided powers -/
+
+open Finset Nat
+
+/-- The divided power structure on an ideal I of a commutative ring A -/
+structure DividedPowers {A : Type*} [CommSemiring A] (I : Ideal A) where
```

Comment:
```suggestion
structure DividedPowers {A : Type*} [Semiring A] (I : Ideal A) where
```
Because why not.

Reply:
I wouldn't do that for a deep mathematical reason: people still do not know what the adequate definition of divided powers on a noncommutative ring should be. I learnt that from Serre who told me about Cartan. Indeed, divided powers were born in the noncommutative world of cohomology of varieties, but Cartan understands the definition only on the even part (which is commutative).
So, while it is likely that there is a usable/useful definition of divided powers on a noncommutative ring, I understand that we don't know what it is yet.

Reply:
OK, no problem.